libroach: detected leaked iterators

petermattis opened the pull request:
Keep track of the number of live iterators and assert when an engine is
closed that the value is 0.

Fixes #23138

Release note: None

petermattis committed:
libroach: detected leaked iterators

Keep track of the number of live iterators and assert when an engine is
closed that the value is 0.

Fixes #23138

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23183)
<!-- Reviewable:end -->

petermattis commented:
It appears that we have a dozen or so tests that leak iterators. Can't merge this PR until I track them all down and fix them.

benesch commented on c-deps/libroach/engine.cc:
```diff
@@ -23,6 +23,15 @@
+DBEngine::~DBEngine() {
+  const int64_t n = iters.load();
+  if (n != 0) {
+    fprintf(stderr, "%lld leaked iterators\n", n);
```
I think you need:

```
fprintf(stderr, "%" PRId64 " leaked iterators\n", n)
```

benesch reviewed (approved):
Code so far LGTM

petermattis commented:
Ok, I figured out all but one of the failures: batches are closed before their iterations are destroyed. This is due to the fast path of closing batches when they commit and is difficult to change. Instead, I moved the code which performs the assertion to only run when the main rocksdb instance is closed, rather than on every close of a snapshot or batch.

`TestImportStmt` is still reporting 12 leaked iterators. I haven't determined if this is real or not. I'm wondering if I should move this iterator leak check from C++ to Go. Doing the check in Go would have the advantage that for race builds we could keep a stack trace to show where the leaked iterator was allocated. Much more difficult to do this from C++. 

Btw, I verified that this assertion caught the leak that Nikhil fixed recently, but it didn't help in tracking it down. Hence my interest in moving the check to Go.

petermattis committed:
ccl/importccl: fix iterator leak

Fix an iterator leak in sstWriter. This one was minor as the associated
rocksdb instance was being closed quickly. The result was a memory leak,
but import is used infrequently so the leak was small.

Release note (bug fix): Fix a small memory leak in IMPORT CSV.

petermattis commented:
Added a second commit which fixes `TestImportStmt`: there was a real leak in the import `sstWriter` code, though the effects were much less serious than previous iterator leaks.

I also added disabled code which tracks the stacks of where iterators are allocated. This is expensive and thus disabled by default, but we can enable if the fast leak check ever fires again.

benesch commented:
This is already paying dividends! How excellent.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 14 of 14 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/engine/rocksdb.go, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6PRK4d6gklXb77vvcB:-L6PRK4d6gklXb77vvcC:beylsfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e567a8a6ed9df3b115317b4abe47ca8a08ac82e/pkg/storage/engine/rocksdb.go#L74)):*
> ```Go
> // not indicate where the leak arose. The expensive checking tracks the stack
> // traces of every iterator allocated. DO NOT ENABLE in production code.
> const debugIteratorLeak = false
> ```

Should we turn this on automatically in race builds? Or is it enough that the assertions are always enabled and we can turn on the stack traces manually?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-:-L6PRX59fbhffVByzX6b:bg7igkv)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 13 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6Riict7jskMr4g7V9f:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think you need:

```
fprintf(stderr, "%" PRId64 " leaked iterators\n", n)
```
</blockquote></details>

Done.

---

*[pkg/storage/engine/rocksdb.go, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6PRK4d6gklXb77vvcB:-L6Rix8TCZfBLSM5RlcN:b-wj35wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e567a8a6ed9df3b115317b4abe47ca8a08ac82e/pkg/storage/engine/rocksdb.go#L74)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should we turn this on automatically in race builds? Or is it enough that the assertions are always enabled and we can turn on the stack traces manually?
</blockquote></details>

It felt sufficient to enable it manually. I'm anxious about making our race builds even slower.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 14 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6S-K55xjFq3He_tNFd:bqlwnk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Using `PRId64` failed when building under the builder. Reverted to `%lld` which we use in at least one other place in `libroach`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*
<!-- Sent from Reviewable.io -->

benesch commented:
How odd. (Your first build using %lld also failed FWIW.)

On Wed, Feb 28, 2018 at 10:26 AM Peter Mattis <notifications@github.com>
wrote:

> Review status: 14 of 15 files reviewed at latest revision, 2 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *c-deps/libroach/engine.cc, line 29 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6S-K55xjFq3He_tNFd:bqlwnk5>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Done.
>
> Using PRId64 failed when building under the builder. Reverted to %lld
> which we use in at least one other place in libroach.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23183>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/23183#issuecomment-369273312>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IGyewUz7uESf5DfGG5RPer-hRY3Oks5tZW9PgaJpZM4SVoQ_>
> .
>

petermattis commented:
> How odd. (Your first build using %lld also failed FWIW.)

The most recent build succeeded and we're definitely using `%lld` elsewhere. What to do?

petermattis commented:
Review status: 12 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[c-deps/libroach/engine.cc, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183#-L6O2G3Mi7St_tBNxLol-r1-29:-L6U71jfdvRSLxLdDgYq:btllbnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/58f1faa27bdf589279155976d7085f918dab55e4/c-deps/libroach/engine.cc#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Using `PRId64` failed when building under the builder. Reverted to `%lld` which we use in at least one other place in `libroach`. 
</blockquote></details>

Back to using `PRId64` (again). How many times can I change this line in one PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23183)*
<!-- Sent from Reviewable.io -->
